clippy_lints/src/indexing_slicing.rs

```diff
@@ -0,0 +1,201 @@
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = [1,2,3,4];
+///
+/// // Bad
+/// x[9];
+/// &x[2..9];
+///
+/// // Good
+/// x[0];
+/// x[3];
+/// ```
+declare_clippy_lint! {
+    pub OUT_OF_BOUNDS_INDEXING,
+    correctness,
+    "out of bounds constant indexing"
+}
+
+/// **What it does:** Checks for usage of indexing or slicing. Does not report
+/// if we can tell that the indexing or slicing operations on an array are in
+/// bounds.
+///
+/// **Why is this bad?** Indexing and slicing can panic at runtime and there are
+/// safe alternatives.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = vec![0; 5];
+/// // Bad
+/// x[2];
+/// &x[2..100];
+/// &x[2..];
+/// &x[..100];
+///
+/// // Good
+/// x.get(2)
+/// x.get(2..100)
+/// x.get(2..)
+/// x.get(..100)
+/// ```
+declare_clippy_lint! {
+    pub INDEXING_SLICING,
+    restriction,
+    "indexing/slicing usage"
+}
+
+#[derive(Copy, Clone)]
+pub struct IndexingSlicingPass;
+
+impl LintPass for IndexingSlicingPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INDEXING_SLICING, OUT_OF_BOUNDS_INDEXING)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicingPass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if let ExprIndex(ref a, ref b) = &expr.node {
+            match &b.node {
+                // Both ExprStruct and ExprPath require this approach's checks
+                // on the `range` returned by `higher::range(cx, b)`.
+                // ExprStruct handles &x[n..m], &x[n..] and &x[..n].
+                // ExprPath handles &x[..] and x[var]
+                ExprStruct(_, _, _) | ExprPath(_) => {
+                    if let Some(range) = higher::range(cx, b) {
+                        let ty = cx.tables.expr_ty(a);
+                        if let ty::TyArray(_, s) = ty.sty {
+                            let size: u128 = s.assert_usize(cx.tcx).unwrap().into();
+                            // Index is a constant range.
+                            if let Some((start, end)) = to_const_range(cx, range, size) {
+                                if start > size || end > size {
+                                    utils::span_lint(
+                                        cx,
+                                        OUT_OF_BOUNDS_INDEXING,
+                                        expr.span,
+                                        "range is out of bounds",
+                                    );
+                                } else {
+                                    // Range is in bounds, ok.
+                                    return;
+                                }
+                            }
+                        }
+                        match (range.start, range.end) {
+                            (None, Some(_)) => {
+                                cx.span_lint(
```

Comment:
Please use the `utils::span_lint()` methods here. These methods add the link to the clippy documentation to the warning message.

There are also methods for help messages in the `utils` module. You could use `utils::span_help_and_lint()` for the `Consider using...` part of the lint.
Bonus points if you use `utils::span_lint_and_sugg()` (Can be automatically fixed by `rustfix`) and generate the right code suggestion. For example: For `x[..3]` the `sugg` argument should be `x.get(..3)`. To do this you have to first figure out, whether `get()` or `get_mut()` is the right function to use and after that extract the span of the range to insert it as the argument for `get[_mut]()`. Something like this: `format!("{}.{}({})", arr/vec_name, "get[_mut]", range.span.as_str())`. 

But this is definitely not necessary and can be done in a later PR. For now `utils::span_help_and_lint()` is enough.

Reply:
Sounds great, thanks!